### PR TITLE
kernel: Add support for transactional systems to update_kernel module

### PR DESCRIPTION
**Pull request by czerw**
- Related ticket: https://progress.opensuse.org/issues/138182
- Needles: none
- Verification run: 
sle:   https://openqa.suse.de/tests/12628830

Verification for incorrect product: http://black-bit.suse.cz/tests/185#step/update_kernel/132
Verification for patch not needed: http://black-bit.suse.cz/tests/187#step/update_kernel/122
Patch verification: http://black-bit.suse.cz/tests/186

### CODE REVIEW

**Comment by mdoucha** [comment]
verification run with `SLE_PRODUCT=slert': https://openqa.suse.de/tests/12563972

File: lib/utils.pm [hunk 1]
```diff
@@ -781,11 +781,15 @@ sub fully_patch_system {
     # Reduce the pressure of serial port when running hyperv test with sle15.
     # poo#115454
     my $zypp_opt = check_var('VIRSH_VMM_FAMILY', 'hyperv') ? '-q' : '';
```
Comment (pevik) [inline lib/utils.pm:783]: `$zypp_opt` is used only for else, maybe move it there?

File: lib/utils.pm [hunk 2]
```diff
@@ -775,17 +775,21 @@ sub fully_patch_system {
         return;
     }
 
-    # Repeatedly call zypper patch until it returns something other than 103 (package manager updates)
     my $ret = 1;
```
Comment (czerw) [inline lib/utils.pm:778]: This is TBD too.

File: tests/kernel/update_kernel.pm [hunk 3]
```diff
@@ -106,7 +112,11 @@ sub update_kernel {
     }
     else {
         # Use single patch or patch list
-        zypper_call("in -l -t patch $patches", exitcode => [0, 102, 103], log => 'zypper.log', timeout => 1400);
+        if (is_transactional) {
+                trup_call("-n -c patch");
```
Comment (czerw) [inline tests/kernel/update_kernel.pm:116]: This section needs improvements as well.

File: tests/kernel/update_kernel.pm [hunk 4]
```diff
@@ -30,7 +31,8 @@ sub check_kernel_package {
     # Only check versioned kernels in livepatch tests. Some old kernel
     # packages install /boot/vmlinux symlink but don't set package ownership.
     my $glob = get_var('KGRAFT', 0) ? '-*' : '*';
-    my $cmd = 'rpm -qf --qf "%{NAME}\n" /boot/vmlinu[xz]' . $glob;
+    my $run_cmd = is_transactional ? 'transactional-update -c -d --quiet run ' : '';
+    my $cmd = $run_cmd . 'rpm -qf --qf "%{NAME}\n" /boot/vmlinu[xz]' . $glob;
```
Comment (czerw) [inline tests/kernel/update_kernel.pm:35]: Not optimal solution, already done in `install _ltp`. Needs separate function.

File: tests/kernel/update_kernel.pm [hunk 5]
```diff
@@ -30,7 +31,8 @@ sub check_kernel_package {
     # Only check versioned kernels in livepatch tests. Some old kernel
```
[the remark this answers is not in the document]
Comment (mdoucha) [reply tests/kernel/update_kernel.pm:31]: And GitHub won't let me comment on a line that's just outside the patch chunk context...

File: lib/utils.pm [hunk 6]
```diff
@@ -781,11 +781,15 @@ sub fully_patch_system {
     # Reduce the pressure of serial port when running hyperv test with sle15.
     # poo#115454
     my $zypp_opt = check_var('VIRSH_VMM_FAMILY', 'hyperv') ? '-q' : '';
-    for (1 .. 3) {
-        $ret = zypper_call("$zypp_opt patch --with-interactive -l", exitcode => [0, 4, 102, 103], timeout => 6000);
-        last if $ret != 103;
+    if (is_transactional) {
+        transactional::trup_call("-c patch");
+        return;
```
[the remark this answers is not in the document]
Comment (czerw) [reply lib/utils.pm:786]: trup patch doesn't rely back proper error code bsc#1216504, decided to do two patch actions  after each other without continue mode

File: tests/kernel/update_kernel.pm (hunk 4 above)
Comment (czerw) [reply tests/kernel/update_kernel.pm]: This modifications are not needed if is system reboot just after the patch. I decided to go with addition reboot to not increase complexity of test code by wrappers.

File: tests/kernel/update_kernel.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (czerw) [reply tests/kernel/update_kernel.pm]: Not needed now, if system is rebooted after the patch. It seems like more correct and safe approach.

File: tests/kernel/update_kernel.pm [hunk 7]
```diff
@@ -106,7 +111,14 @@ sub update_kernel {
...
         # Use single patch or patch list
-        zypper_call("in -l -t patch $patches", exitcode => [0, 102, 103], log => 'zypper.log', timeout => 1400);
+        if (is_transactional) {
+            # Proceed with transactional-update patch
+            trup_call("patch");
+            # Reboot system after patch, to make sure that further checks are done on updated system
+            reboot_on_changes;
+        } else {
+            zypper_call("in -l -t patch $patches", exitcode => [0, 102, 103], log => 'zypper.log', timeout => 1400);
+        }
```
Comment (foursixnine) [inline tests/kernel/update_kernel.pm:121]: I guess instances of this is where the follow ups to https://progress.opensuse.org/issues/117028 are to be added, right?

something to note would be that `patch` is also something that has to be there too.

@fgerling Can you add a ticket to our backlog about this?

**Review by mdoucha — APPROVED** [review]
Looks like SLE-Micro has no `kernel-devel` packages at all, not even for `kernel-default`. LGTM then.

File: tests/kernel/update_kernel.pm (hunk 7 above)
Comment (czerw) [reply tests/kernel/update_kernel.pm:121]: > something to note would be that `patch` is also something that has to be there too.

I made comment in  mentioned ticket regarding `patch`.

**Review by pevik — APPROVED** [review]
LGTM, but there are others more familiar with Alp.

**Pull request merged**